platoon/channel/worker.py

```diff
@@ -0,0 +1,569 @@
+        else:
+            raise AttributeError("pygpu or theano is not imported")
+
+    def init_mb_sock(self, port, hwm=10):
+        """
+        Initialize the mini-batch socket.
+
+        This must be called before using :meth:`recv_mb`.
+
+        Parameters
+        ----------
+        port : int
+            The port to reach the mini-batch server on.
+        hwm : int
+            High water mark, see pyzmq docs.
+
+        """
+        self.asocket = self.context.socket(zmq.PULL)
+        self.asocket.setsockopt(zmq.LINGER, 0)
+        self.asocket.set_hwm(hwm)
+        self.asocket.connect("tcp://localhost:{}".format(port))
+
+        self.apoller = zmq.Poller()
+        self.apoller.register(self.asocket, zmq.POLLIN)
+
+    def _init_control_socket(self, port):
+        """
+        Initialize control socket.
+
+        This must be called before using :meth:`send_req`.
+
+        Parameters
+        ---------
+        port : int
+            Port where the control master is listening on.
+
+        """
+        self.csocket = self.context.socket(zmq.REQ)
+        self.csocket.setsockopt(zmq.LINGER, 0)
+        self.csocket.connect('tcp://localhost:{}'.format(port))
+
+        self.cpoller = zmq.Poller()
+        self.cpoller.register(self.csocket, zmq.POLLIN)
+
+################################################################################
+#                            Collectives Interface                             #
+################################################################################
+
+    def shared(self, array):
+        """Creates a new posix shared memory buffer to be shared among Workers
+        and their Controller and maps the size of `array` to that buffer.
+
+        Controller is requested to create a new shared memory buffer with the
+        same size as `array` in order to be used in multi-node/gpu platoon
+        collective operations through the new worker interface. All participants
+        in a host have access to that memory.
+
+        :param array: will correspond to shared buffer in host with the same size
+        :type array: :ref:`pygpu.gpuarray.GpuArray`
+
+        Notes
+        -----
+        For internals: There should probably be a barrier across hosts' worker
+        here to ensure that so far, controller has serviced all workers (if
+        needed) a new shared memory's name. This is due to the fact that
+        Controller can service one Worker at a time and a platoon collective
+        service is a blocking one across Controllers. Current implementation
+        is valid because calls to pygpu.collectives interface are synchronous
+        across workers.
+
+        """
+        if not isinstance(array, pygpu.gpuarray.GpuArray):
+            raise TypeError("`array` input is not pygpu.gpuarray.GpuArray.")
+
+        bytesize = array.size * array.itemsize
+
+        if bytesize in self.shared_arrays:
+            return self.shared_arrays[bytesize]
```

Comment:
This is not a problem, unless we have concurrent calls in **all_reduce** in the same worker-process and we are running in multi-node. But still it is quite a hack. I will describe the coding history so that you can help me find a better solution:

1. Initially, I thought of mapping a destination `pygpu.gpuarray.GpuArray` to a shared memory buffer of the same bytesize. Having a buffer in host with the same size is a necessity, in order to execute the collective in multi-node, because so far this operation can be only executed using hosts' CPUs. Having a different shared buffer corresponding to each `GpuArray` was a naive and safe solution. But, `pygpu.gpuarray.GpuArray` is not hashable and I did not want to export a key that the user needed to keep in mind that it corresponds with a specific `GpuArray`. The latter would make also the API not clean because it would look like `Worker.all_reduce(self, src, op, dest=None, dest_id=None)`... The key needed to be inferred from or self contained in the destination object.

2. So I thought it was a good idea to use `theano.gpuarray.GpuArraySharedVariable` in the API, which are hashable and do contain the `pygpu.gpuarray.GpuArray`. Getting into the underlying GPU buffer is a `array.get_value(borrow=True, return_internal_type=True)`. But this was about to make a mess afterwards, that I did not expect when I tested the interface. Trying to implement with this new interface a synchronous SGD in as a GlobalDynamics class impl, I found that mixing Theano with this new interface was not a good idea. Returning a single function that would execute the AverageSGD would have looked like this:
    ```python
    def make_rule(self, local_updates):
        import theano
        global_updates = []
        for update in local_updates:
            global_updates.append(update / self.worker.global_size)
        after_fn = theano.function([], [],
            updates=list(zip(local_updates, global_updates)),
            accept_inplace = True)
        def fn():
            for update in local_updates:
                self.worker.all_reduce(lup, '+', lup)
            after_fn()
        return fn
    ```
Yuck. If mathematical operations are needed before and after the `all_reduce`?..
So I decided to implement a `theano.Op` to wrap the `Worker.all_reduce` interface. While doing so, I realized that interface needed to return back to using `pygpu.gpuarray.GpuArray`, because hacking through `node` argument in `AllReduce.perform` was not viable. If theano operations were preceding the AllReduce{...} Op then the input variables to that Apply node would be `theano.gpuarray.GpuArrayVariable`s and not `GpuArraySharedVariables` and thus not able to call the Worker interface.

3. I realized that shared memory buffer is truly temporary in functionality and the reason that we are not unlinking and creating it again in every call is for performance reasons. So as its usage is only within `Worker.all_reduce` and out of this call it does not need to contain something, I thought that it can be reused by calls that have the same bytesize. And thus I chose to have bytesize as a key.